### PR TITLE
mrc-2795 create structured generic logging

### PR DESCRIPTION
**Pull request by LekanAnni**
This PR logs structured data which gets shipped to ELK in Json format. 

My solution uses spring-boot logging implementation, which was previously disabled. I believe disabling spring-logging was because slf4j-simple was implemented within postgreSQL which was conflicting with spring logging framework. I upgraded Postgres to the latest version which has now dropped slf4j-simple logging as part of it's dependencies.

Then,  I enabled spring-logging as hint default logging so as to configure logging behaviour.  Config allows  for formatting of logs in Json. The config file `logback-spring.xml`  renders  logs to console as `STDOUT` when in development environment and as Json when not in dev environment.

The intermittent failure has resurfaced again, once we restart the build kite server, everything should be fine. While I have not found a real solution to the stale buildkite state, I think manual reboot fixes the issue for now.

To test normal format:
*  Run the app
*  go to project history page
* edit project note
* then check the logs display normally in app console 

To test Json format
* Change Spring profile in grade.properties 
* follow the above test
* then check logs display in Json

Please speak to Rob to gain access to ELK to see display

A simple visualisation dashboard showing the test log

Screenshot 2022-05-19 at 14 59 07

link to the dashboard, you need to be on side network.
https://logs.dide.ic.ac.uk/app/dashboards#/view/68f9cd60-679d-11eb-9ee9-77ae80d7fb74?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:'2022-05-19T09:32:08.152Z',to:'2022-05-28T10:02:25.502Z'))

For a better visualisation, we need to create a mock up for visualisation, deciding what to see on hint dashboard.

## Type of version change
_Delete as appropriate_

Major - Minor - Patches - None

## Checklist

- [x] I have incremented version number, or version needs no increment
- [x] The build passed successfully, or failed because of ADR tests

### CODE REVIEW

File: src/app/src/test/kotlin/org/imperial/mrc/hint/integration/ExceptionHandlerTests.kt
Comment (EmmaLRussell) [inline src/app/src/test/kotlin/org/imperial/mrc/hint/integration/ExceptionHandlerTests.kt:159]: Why was this change needed?

File: src/app/src/test/kotlin/org/imperial/mrc/hint/unit/logging/LoggingTests.kt
Comment (EmmaLRussell) [inline src/app/src/test/kotlin/org/imperial/mrc/hint/unit/logging/LoggingTests.kt:9]: Could we call this class and file GenericLoggerImplTests if that's what it's testing? That would follow our standard pattern.

File: src/app/src/main/kotlin/org/imperial/mrc/hint/logging/GenericLoggerImpl.kt
Comment (EmmaLRussell) [inline src/app/src/main/kotlin/org/imperial/mrc/hint/logging/GenericLoggerImpl.kt:14]: This "TEST INFO" is meant to be a format string, according to Logger docs - what's the plan for this (and for the error case too)?

**Review by Nicolas-Dolan — COMMENTED** [review]
LGTM. Just a few code coverage issues to be addressed. Any advice on how I might manually test this? What should I be looking for?

**Comment by LekanAnni** [comment]
> LGTM. Just a few code coverage issues to be addressed. Any advice on how I might manually test this? What should I be looking for?

I have updated the description

**Review by EmmaLRussell — COMMENTED** [review]
This seems like a nice approach, but I'm a bit confused about where the json is created. This should be generated when the logger is using the json console appender (in non-dev), from the `kv` parameters provided to `info` and `error`? Have you actually seen the json generated? Would it be possible to automate a test for that?

In dev mode, all that gets logged is: `2022-04-05 11:17:47.785  INFO 26499 --- [nio-8080-exec-5] o.i.mrc.hint.logging.GenericLoggerImpl   : TEST INFO`  - where TEST INFO is the format string that's been given to the logger method. It would be nice if the logging data could be logged to the console too. 

Will we continue to get standard logging to the console in the docker containers in non-dev? I think we definitely want to retain that for debugging. 

The ticket mentions making a simple visualisation. Were you able to actually ship any logs to ELK to do this? Do we ship logs from staging? Could do it from there?

Should get the build through once mrc-3111 is merged, hopefully. Seem to be a couple of codecov grumbles.

File: src/app/src/main/kotlin/org/imperial/mrc/hint/logging/GenericLoggerImpl.kt
Comment (LekanAnni) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/logging/GenericLoggerImpl.kt:14]: This is more like a placeholder for this MVP but I have added a more reasonable message.

File: src/app/src/test/kotlin/org/imperial/mrc/hint/integration/ExceptionHandlerTests.kt
Comment (LekanAnni) [reply src/app/src/test/kotlin/org/imperial/mrc/hint/integration/ExceptionHandlerTests.kt:159]: Yes, I updated Postgres. The newer version does not support mocking for PSQLState.

**Comment by LekanAnni** [comment]
> LGTM. Just a few code coverage issues to be addressed. Any advice on how I might manually test this? What should I be looking for?

Console should log a structured message with serialised json.

**Comment by LekanAnni** [comment]
Please do not merge, I am in the process of creating visualisation in Kibana.

File: src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ProjectsController.kt
Comment (EmmaLRussell) [inline src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ProjectsController.kt:107]: If this is your example case, you'll probably want to include some tags too e.g. `tags=listOf("project","notes")`

**Review by Nicolas-Dolan — COMMENTED** [review]
I tried testing manually and it seems to be working ok, but I'm not sure what is meant by "logs display normally". Should I be expecting a certain behaviour? I don't have much to contribute review-wise as I think I would need you to take me through what you are doing for me to properly understand it.

File: src/app/src/main/kotlin/org/imperial/mrc/hint/Annotations.kt
Comment (LekanAnni) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/Annotations.kt:6]: Data classes are tested in the parent class i.e LogMetadata. Children data classes are not used directly and are just for serialisation. In fact these children classes are tested in GenericLoggerImplTests. Code coverage was complaining because they had been tested within LogMetadata.

File: src/app/src/main/kotlin/org/imperial/mrc/hint/logging/GenericLoggerImpl.kt
Comment (LekanAnni) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/logging/GenericLoggerImpl.kt:25]: I understand we have a decision to make about this formatting and all, I don't think this ticket is meant to address all our concerns.

**Comment by LekanAnni** [comment]
> I tried testing manually and it seems to be working ok, but I'm not sure what is meant by "logs display normally". Should I be expecting a certain behaviour? I don't have much to contribute review-wise as I think I would need you to take me through what you are doing for me to properly understand it.

I meant the logs are sent to STDOUT which can be seen in console. You should see logs in json format. I will have more explanation as soon as Kibana dashboard is ready.

File: src/app/src/main/kotlin/org/imperial/mrc/hint/logging/LogMetadata.kt
Comment (LekanAnni) [inline src/app/src/main/kotlin/org/imperial/mrc/hint/logging/LogMetadata.kt:39]: I have just added NoCoverage annotation to LogMetadata because code coverage now complains after verifying the data object rather than verifying each object properties.

File: src/app/src/main/kotlin/org/imperial/mrc/hint/logging/GenericLoggerImpl.kt
Comment (EmmaLRussell) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/logging/GenericLoggerImpl.kt]: Commented above on this - added a ticket for this issue: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-3250

**Comment by EmmaLRussell** [comment]
> > I tried testing manually and it seems to be working ok, but I'm not sure what is meant by "logs display normally". Should I be expecting a certain behaviour? I don't have much to contribute review-wise as I think I would need you to take me through what you are doing for me to properly understand it.
> 
> I meant the logs are sent to STDOUT which can be seen in console. You should see logs in json format. I will have more explanation as soon as Kibana dashboard is ready.

This isn't the case as far as I can tell (as discussed above) and probably not what we want anyway, as the json logs aren't going to be very human readable.

File: src/app/src/main/kotlin/org/imperial/mrc/hint/logging/GenericLoggerImpl.kt
Comment (EmmaLRussell) [inline src/app/src/main/kotlin/org/imperial/mrc/hint/logging/GenericLoggerImpl.kt:16]: This has changed since the last iteration - don't need to include all values as they get serialized automatically?

**Review by EmmaLRussell — COMMENTED** [review]
Still think I'll need the logs via kibana to be sure I understand how it's all working. Think there might be some crossed wires about whether JSON should be logged to console or not - that's not what I'm seeing (see below), and I thought the 'msg' parameter in the logger defined what would be logged to console. 
image

File: src/app/src/main/kotlin/org/imperial/mrc/hint/Annotations.kt
Comment (EmmaLRussell) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/Annotations.kt:6]: I just noticed we do have this in OW, so I guess there's precedent (but I probably complained about that too! 😆) - I think that is a slightly different use case as that's all for types passed through from Orderly Server rather than objects created and used internally.

Comment (LekanAnni) [reply src/app/src/main/kotlin/org/imperial/mrc/hint/Annotations.kt:6]: I have now removed noCoverage annotations.

**Comment by LekanAnni** [comment]
> Still think I'll need the logs via kibana to be sure I understand how it's all working. Think there might be some crossed wires about whether JSON should be logged to console or not - that's not what I'm seeing (see below), and I thought the 'msg' parameter in the logger defined what would be logged to console.
> image

This issue has now been looked at, same log messages will be printed for dev and non-dev. non-dev will be in json while dev will log only it's message as json.

**Comment by LekanAnni** [comment]
> > This issue has now been looked at, same log messages will be printed for dev and non-dev. non-dev will be in json while dev will log only it's message as json.
> 
> OK, nice - I can now see the actual object being logged locally (not in JSON, but I guess that's java-stringified).
> 
> image And then if I run it in non-dev (setting `springProfile=prod` in gradle.properties), I can see that the whole log message is now in JSON, including the logMetadata parts under the 'hint' property (although the non-JSON "message" is still in there as a separate field (but I guess that's ok):
> 
> ```
> {
>   "@timestamp": "2022-06-09T10:35:39.55+01:00",
>   "@version": "1",
>   "message": "hint=LogMetadata(username=test.user@example.com, app=AppOrigin(name=hint, type=backend), request=Request(method=POST, path=/project/8/note, hostname=localhost, client=Client(agent=Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36, geoIp=0:0:0:0:0:0:0:1, sessionId=33865022C6D0BC727F0508D1FFFF781A)), response=null, error=null, action=Updating project note, tags=[project, notes])",
>   "logger_name": "org.imperial.mrc.hint.logging.GenericLoggerImpl",
>   "thread_name": "http-nio-8080-exec-3",
>   "level": "INFO",
>   "level_value": 20000,
>   "hint": {
>     "username": "test.user@example.com",
>     "app": {
>       "name": "hint",
>       "type": "backend"
>     },
>     "request": {
>       "method": "POST",
>       "path": "/project/8/note",
>       "hostname": "localhost",
>       "client": {
>         "agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.82 Safari/537.36",
>         "geoIp": "0:0:0:0:0:0:0:1",
>         "sessionId": "33865022C6D0BC727F0508D1FFFF781A"
>       }
>     },
>     "response": null,
>     "error": null,
>     "action": "Updating project note",
>     "tags": [
>       "project",
>       "notes"
>     ]
>   }
> }
> ```
> 
> This seems good to me. Is there a kibana report to look at yet?

Visualisation is created in table formats, as mentioned above, we need to determine what to create when we gather adequate data. However, see below  link for table view on Kibana https://logs.dide.ic.ac.uk/app/dashboards#/view/68f9cd60-679d-11eb-9ee9-77ae80d7fb74?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:'2022-05-19T09:32:08.152Z',to:'2022-05-28T10:02:25.502Z'))

**Comment by EmmaLRussell** [comment]
> Visualisation is created in table formats, as mentioned above, we need to determine what to create when we gather adequate data. However, see below link for table view on Kibana https://logs.dide.ic.ac.uk/app/dashboards#/view/68f9cd60-679d-11eb-9ee9-77ae80d7fb74?_g=(filters:!(),refreshInterval:(pause:!t,value:0),time:(from:'2022-05-19T09:32:08.152Z',to:'2022-05-28T10:02:25.502Z'))

Nice! :sunglasses:

**Review by EmmaLRussell — APPROVED** [review]

**Review by Nicolas-Dolan — APPROVED** [review]
LGTM

**Pull request merged**
